### PR TITLE
Add a batch size for cleanup

### PR DESCRIPTION
**Pull request by eyurtsev**
Add pagination to indexing cleanup to deal with large numbers of documents that
need to be deleted.

### CODE REVIEW

File: libs/langchain/langchain/indexes/_api.py [hunk 1]
```diff
@@ -329,14 +331,20 @@ def index(
...
+
+            if not uids_to_delete:
+                break
+
+            if uids_to_delete:
+                # First delete from record store.
+                vector_store.delete(uids_to_delete)
+                # Then delete from record manager.
+                record_manager.delete_keys(uids_to_delete)
+                num_deleted += len(uids_to_delete)
```
Comment (baskaryan) [inline libs/langchain/langchain/indexes/_api.py:347]: nit / up to you
```suggestion
        uids_to_delete = record_manager.list_keys(
            before=index_start_dt, limit=cleanup_batch_size
        )
        while uids_to_delete:
              # First delete from record store.
              vector_store.delete(uids_to_delete)
              # Then delete from record manager.
              record_manager.delete_keys(uids_to_delete)
              num_deleted += len(uids_to_delete)
              
              uids_to_delete = record_manager.list_keys(
                  before=index_start_dt, limit=cleanup_batch_size
              )

```

Comment (baskaryan) [reply libs/langchain/langchain/indexes/_api.py:347]: could even:
```python
while uids_to_delete := record_manager...
```

File: libs/langchain/tests/unit_tests/indexes/test_indexing.py [hunk 2]
```diff
@@ -472,3 +472,42 @@ def test_deduplication(
...
+    """Check that we can clean up with different batch size."""
+    docs = [
+        Document(
+            page_content="This is a test document.",
+            metadata={"source": str(d)},
+        )
+        for d in range(1000)
+    ]
+
+    # Should result in only a single document being added
```
Comment (baskaryan) [inline libs/langchain/tests/unit_tests/indexes/test_indexing.py:489]: comment

File: libs/langchain/langchain/indexes/_sql_record_manager.py [hunk 3]
```diff
@@ -259,6 +259,7 @@ def list_keys(
         before: Optional[float] = None,
         after: Optional[float] = None,
         group_ids: Optional[Sequence[str]] = None,
+        limit: Optional[int] = None,
```
Comment (baskaryan) [inline libs/langchain/langchain/indexes/_sql_record_manager.py:262]: should there be a pagination option too? is that easy? can be for another pr, just wondering

Comment (eyurtsev) [reply libs/langchain/langchain/indexes/_sql_record_manager.py:262]: We don't need offset right now, so didn't want to introduce it since offset also implies that we need to sort (which we don't want)

**Comment by baskaryan** [comment]
any idea if the mypy error is relevant?

File: libs/langchain/langchain/indexes/_api.py (hunk 1 above)
Comment (eyurtsev) [reply libs/langchain/langchain/indexes/_api.py:347]: Isn't the `while True` more explicit about the fact that there is a potential for an infinite loop? I don't mind either way though

**Review by baskaryan — APPROVED** [review]
